### ISSUE TITLE
`blas.ldflags` doesn't work as advertised

### ISSUE THREAD

**Issue by zym1010** [Theano/Theano#4688]
My python 2.7 environment has numpy and scipy installed from conda-forge, which only offers single-threaded OpenBLAS. Therefore, to use Theano efficiently, I compiled a multithread version of OpenBLAS, and set `.theanorc` as follows.

~~~
[global]
device = cpu
floatX = float32
[blas]
ldflags = -L/home/yimengzh_everyday/lib/OpenBLAS/lib -lopenblas
~~~

However, when doing BLAS test in Theano, clearly only one thread is being used.

To make sure this is not a problem of name confusion since openblas is available both from python's `lib` directory as well as `/home/yimengzh_everyday/lib/OpenBLAS/lib`, I renamed all files in `/home/yimengzh_everyday/lib/OpenBLAS/lib` to `libopenblasp`, so I have `-lopenblasp` in `ldflags` to avoid potential name collision.

However, the problem persists. I checked the compiled program in `~/.theano`, and, to my surprise, it's linked against `libopenblas`, not the `libopenblasp`.

~~~
$ ldd tmpHL25Of/6fc8b974e3e5b266cb7c150fc09bc91c.so
	linux-vdso.so.1 =>  (0x00007ffff35fc000)
	libopenblas.so.0 => not found
	libpython2.7.so.1.0 => /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0 (0x00007f7994039000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f7993d35000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f7993970000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f7993752000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f7993539000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f7993335000)
	libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007f7993132000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f7992e2c000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f79947a3000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f7992c16000)
~~~

This seems to be the case for both a CentOS 6.2 machine, as well as a Ubuntu 14.04 one.

**Comment by nouiz** [comment]
This work well for us. Do you have environment variable like
OMP_NUM_THREADS=1 or *_NUM_THREADS=1? This would tell openblas to use only
1 thread.

How was your openblas installed? Depending of how it was compiled, it could
be compiled with only 1 thread support.

How do you see it use only 1 thread? Some pre-compiled version are limited
to 2 threads.

If the matrix size are too small, more then 1 thread won't help, make sure
to test with big matrix size.

**Comment by zym1010** [comment]
@nouiz 

> This work well for us. Do you have environment variable like
OMP_NUM_THREADS=1 or *_NUM_THREADS=1? This would tell openblas to use only
1 thread.

Nope. I can read that all these flags are `None` based on your test blas script.

> How was your openblas installed? Depending of how it was compiled, it could
be compiled with only 1 thread support.

To make sure that there's no problem with my installation, I copied my own blas libraries to overwrite the versions `numpy` linked to, and then the running time dropped from 40s to 5s.

Basically, initially, my own blas is multithread, and numpy's blas is single thread.

When I numpy's blas is single-thread, no matter how I play with ldflags, I can only get single thread.
When I replaced numpy's blas with my own multi-thread version, even when I link my ldflags to a single-thread version, still I can multithread version.

**Comment by zym1010** [comment]
When using single thread OpenBLAS from numpy

~~~
Some Theano flags:
    blas.ldflags= -L/home/yimengzh_everyday/miniconda2-test/envs/theano-tf/lib -lopenblas -lopenblas -Wl,-rpath,/home/yimengzh_everyday/miniconda2-test/envs/theano-tf/lib
    compiledir= /home/yimengzh_everyday/.theano/compiledir_Linux-3.13--generic-x86_64-with-debian-jessie-sid-x86_64-2.7.12-64
    floatX= float32
    device= cpu
Some OS information:
    sys.platform= linux2
    sys.version= 2.7.12 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:42:40)
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]
    sys.prefix= /home/yimengzh_everyday/miniconda2-test/envs/theano-tf
Some environment variables:
    MKL_NUM_THREADS= None
    OMP_NUM_THREADS= None
    GOTO_NUM_THREADS= None

Numpy config: (used when the Theano flag "blas.ldflags" is empty)
lapack_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/home/yimengzh_everyday/miniconda2-test/envs/theano-tf/lib']
    define_macros = [('HAVE_CBLAS', None)]
    language = c
blas_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/home/yimengzh_everyday/miniconda2-test/envs/theano-tf/lib']
    define_macros = [('HAVE_CBLAS', None)]
    language = c
openblas_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/home/yimengzh_everyday/miniconda2-test/envs/theano-tf/lib']
    define_macros = [('HAVE_CBLAS', None)]
    language = c
openblas_lapack_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/home/yimengzh_everyday/miniconda2-test/envs/theano-tf/lib']
    define_macros = [('HAVE_CBLAS', None)]
    language = c
blas_mkl_info:
  NOT AVAILABLE
Numpy dot module: numpy.core.multiarray
Numpy location: /home/yimengzh_everyday/miniconda2-test/envs/theano-tf/lib/python2.7/site-packages/numpy/__init__.pyc
Numpy version: 1.11.0

We executed 10 calls to gemm with a and b matrices of shapes (5000, 5000) and (5000, 5000).

Total execution time: 32.50s on CPU (with direct Theano binding to blas).

Try to run this script a few times. Experience shows that the first time is not as fast as followings calls. The difference is not big, but consistent.
~~~

When setting ldflags to my own openblas

~~~
Some Theano flags:
    blas.ldflags= -L/home/yimengzh_everyday/lib/OpenBLAS/lib -lopenblasp
    compiledir= /home/yimengzh_everyday/.theano/compiledir_Linux-3.13--generic-x86_64-with-debian-jessie-sid-x86_64-2.7.12-64
    floatX= float32
    device= cpu
Some OS information:
    sys.platform= linux2
    sys.version= 2.7.12 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:42:40)
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]
    sys.prefix= /home/yimengzh_everyday/miniconda2-test/envs/theano-tf
Some environment variables:
    MKL_NUM_THREADS= None
    OMP_NUM_THREADS= None
    GOTO_NUM_THREADS= None

Numpy config: (used when the Theano flag "blas.ldflags" is empty)
lapack_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/home/yimengzh_everyday/miniconda2-test/envs/theano-tf/lib']
    define_macros = [('HAVE_CBLAS', None)]
    language = c
blas_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/home/yimengzh_everyday/miniconda2-test/envs/theano-tf/lib']
    define_macros = [('HAVE_CBLAS', None)]
    language = c
openblas_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/home/yimengzh_everyday/miniconda2-test/envs/theano-tf/lib']
    define_macros = [('HAVE_CBLAS', None)]
    language = c
openblas_lapack_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/home/yimengzh_everyday/miniconda2-test/envs/theano-tf/lib']
    define_macros = [('HAVE_CBLAS', None)]
    language = c
blas_mkl_info:
  NOT AVAILABLE
Numpy dot module: numpy.core.multiarray
Numpy location: /home/yimengzh_everyday/miniconda2-test/envs/theano-tf/lib/python2.7/site-packages/numpy/__init__.pyc
Numpy version: 1.11.0

We executed 10 calls to gemm with a and b matrices of shapes (5000, 5000) and (5000, 5000).

Total execution time: 32.85s on CPU (with direct Theano binding to blas).

Try to run this script a few times. Experience shows that the first time is not as fast as followings calls. The difference is not big, but consistent.
~~~

When copying my own blas to numpy directory:

~~~
(theano-tf)yimengzh_everyday@leelabgpu:~$ A_FILE_NEW=/home/yimengzh_everyday/lib/OpenBLAS/lib/libopenblasp.a
(theano-tf)yimengzh_everyday@leelabgpu:~$ SO_FILE_NEW=/home/yimengzh_everyday/lib/OpenBLAS/lib/libopenblasp.so
(theano-tf)yimengzh_everyday@leelabgpu:~$ A_FILE=/home/yimengzh_everyday/miniconda2-test/envs/theano-tf/lib/libopenblas-r0.2.18.a
(theano-tf)yimengzh_everyday@leelabgpu:~$ SO_FILE=/home/yimengzh_everyday/miniconda2-test/envs/theano-tf/lib/libopenblas-r0.2.18.so
(theano-tf)yimengzh_everyday@leelabgpu:~$ cp ${A_FILE_NEW} ${A_FILE}
(theano-tf)yimengzh_everyday@leelabgpu:~$ cp ${SO_FILE_NEW} ${SO_FILE}
~~~

Then the result is totally different.

~~~
Some Theano flags:
    blas.ldflags= -L/home/yimengzh_everyday/lib/OpenBLAS/lib -lopenblasp
    compiledir= /home/yimengzh_everyday/.theano/compiledir_Linux-3.13--generic-x86_64-with-debian-jessie-sid-x86_64-2.7.12-64
    floatX= float32
    device= cpu
Some OS information:
    sys.platform= linux2
    sys.version= 2.7.12 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:42:40)
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]
    sys.prefix= /home/yimengzh_everyday/miniconda2-test/envs/theano-tf
Some environment variables:
    MKL_NUM_THREADS= None
    OMP_NUM_THREADS= None
    GOTO_NUM_THREADS= None

Numpy config: (used when the Theano flag "blas.ldflags" is empty)
lapack_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/home/yimengzh_everyday/miniconda2-test/envs/theano-tf/lib']
    define_macros = [('HAVE_CBLAS', None)]
    language = c
blas_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/home/yimengzh_everyday/miniconda2-test/envs/theano-tf/lib']
    define_macros = [('HAVE_CBLAS', None)]
    language = c
openblas_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/home/yimengzh_everyday/miniconda2-test/envs/theano-tf/lib']
    define_macros = [('HAVE_CBLAS', None)]
    language = c
openblas_lapack_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/home/yimengzh_everyday/miniconda2-test/envs/theano-tf/lib']
    define_macros = [('HAVE_CBLAS', None)]
    language = c
blas_mkl_info:
  NOT AVAILABLE
Numpy dot module: numpy.core.multiarray
Numpy location: /home/yimengzh_everyday/miniconda2-test/envs/theano-tf/lib/python2.7/site-packages/numpy/__init__.pyc
Numpy version: 1.11.0

We executed 10 calls to gemm with a and b matrices of shapes (5000, 5000) and (5000, 5000).

Total execution time: 4.22s on CPU (with direct Theano binding to blas).

Try to run this script a few times. Experience shows that the first time is not as fast as followings calls. The difference is not big, but consistent
~~~
